lightning_examples/barlow-twins/barlow_twins.py

```diff
@@ -0,0 +1,569 @@
+
+# %%
+def cifar10_normalization():
+    normalize = transforms.Normalize(
+        mean=[x / 255.0 for x in [125.3, 123.0, 113.9]], std=[x / 255.0 for x in [63.0, 62.1, 66.7]]
+    )
+    return normalize
+
+
+train_transform = BarlowTwinsTransform(
+    input_height=32, gaussian_blur=False, jitter_strength=0.5, normalize=cifar10_normalization()
+)
+train_dataset = CIFAR10(root=".", train=True, download=True, transform=train_transform)
+
+val_transform = BarlowTwinsTransform(
+    input_height=32, gaussian_blur=False, jitter_strength=0.5, normalize=cifar10_normalization()
+)
+val_dataset = CIFAR10(root=".", train=False, download=True, transform=train_transform)
+
+train_loader = DataLoader(
+    train_dataset, batch_size=batch_size, shuffle=True, num_workers=num_workers, drop_last=True, pin_memory=True
+)
+val_loader = DataLoader(
+    val_dataset, batch_size=batch_size, shuffle=False, num_workers=num_workers, drop_last=True, pin_memory=True
+)
+
+
+# %% [markdown]
+# ### Barlow Twins Loss
+#
+# Here we define the loss function for Barlow Twins. It first normalizes the D dimensinonal vectors from the projection head and then computes the DxD cross-correlation matrix between the normalized vectors of the 2 views of each image.
+#
+# Then it splits this cross-correlation matrix into two parts. The first part, the diagonal of this matrix is brought closer to 1, which pushes up the cosine similarity between the latent vectors of two views of each image, thus making the backbone invariant to the transformations applied to the views. The second part of the loss pushes the non-diagonal elements of the cross-corrlelation matrix closes to 0. This reduces the redundancy between the different dimensions of the latent vector.
+
+# %%
+class BarlowTwinsLoss(nn.Module):
+    def __init__(self, batch_size, lambda_coeff=5e-3, z_dim=128):
+        super().__init__()
+
+        self.z_dim = z_dim
+        self.batch_size = batch_size
+        self.lambda_coeff = lambda_coeff
+
+    def off_diagonal_ele(self, x):
+        # taken from: https://github.com/facebookresearch/barlowtwins/blob/main/main.py
+        # return a flattened view of the off-diagonal elements of a square matrix
+        n, m = x.shape
+        assert n == m
+        return x.flatten()[:-1].view(n - 1, n + 1)[:, 1:].flatten()
+
+    def forward(self, z1, z2):
+        # N x D, where N is the batch size and D is output dim of projection head
+        z1_norm = (z1 - torch.mean(z1, dim=0)) / torch.std(z1, dim=0)
+        z2_norm = (z2 - torch.mean(z2, dim=0)) / torch.std(z2, dim=0)
+
+        cross_corr = torch.matmul(z1_norm.T, z2_norm) / self.batch_size
+
+        on_diag = torch.diagonal(cross_corr).add_(-1).pow_(2).sum()
+        off_diag = self.off_diagonal_ele(cross_corr).pow_(2).sum()
```

Comment:
Might be cleaner, faster to multiple this with a mask instead. (cross_corr * (1 - torch.eye(self.batch_size)).pow_(2).sum()

Reply:
Keeping this consistent with the official implementation: https://github.com/facebookresearch/barlowtwins/blob/main/main.py#L218